src/ai/backend/manager/models/etcd.py

```diff
@@ -0,0 +1,257 @@
+        }
+        container_registry_iv.check(raw_config)
+        config = {key: value for key, value in raw_config.items() if value is not None}
+
+        log.info(
+            "ETCD.CREATE_CONTAINER_REGISTRY (ak:{}, key: {}, hostname:{}, config:{})",
+            ctx.access_key,
+            ETCD_CONTAINER_REGISTRY_KEY,
+            hostname,
+            config,
+        )
+        updates = {}
+
+        def flatten(o):
+            for k, v in o.items():
+                if k == "":
+                    inner_prefix = f"{ETCD_CONTAINER_REGISTRY_KEY}/{hostname}"
+                else:
+                    inner_prefix = f"{ETCD_CONTAINER_REGISTRY_KEY}/{hostname}/{k}"
+                if isinstance(v, Mapping):
+                    flatten(v)
+                else:
+                    updates[inner_prefix] = v
+
+        flatten(config)
+        # TODO: chunk support if there are too many keys
+        if len(updates) > 16:
+            raise InvalidAPIParameters("Too large update! Split into smaller key-value pair sets.")
+        await ctx.shared_config.etcd.put_dict(updates)
+        return cls(result="ok")
+
+
+class ModifyContainerRegistry(graphene.Mutation):
+    allowed_roles = (UserRole.SUPERADMIN,)
+    result = graphene.String()
+
+    class Arguments:
+        hostname = graphene.String(required=True)
+        props = ModifyContainerRegistryInput(required=True)
+
+    @classmethod
+    @privileged_mutation(
+        UserRole.SUPERADMIN,
+        lambda id, **kwargs: (None, id),
+    )
+    async def mutate(
+        cls, root, info: graphene.ResolveInfo, hostname: str, props: CreateContainerRegistryInput
+    ) -> CreateContainerRegistry:
+        ctx: GraphQueryContext = info.context
+        if await ctx.shared_config.etcd.get(f"{ETCD_CONTAINER_REGISTRY_KEY}/{hostname}") is None:
+            raise ObjectNotFound(object_name=f"registry: {hostname}")
+
+        raw_config = {
+            "": props.url,
+            "type": props.container_type,
+            "project": props.project,
+            "username": props.username,
+            "password": props.password,
+        }
+        container_registry_iv.check(raw_config)
+        config = {key: value for key, value in raw_config.items() if value is not None}
+
+        log.info(
+            "ETCD.MODIFY_CONTAINER_REGISTRY (ak:{}, key: {}, hostname:{}, config:{})",
+            ctx.access_key,
+            ETCD_CONTAINER_REGISTRY_KEY,
+            hostname,
+            config,
+        )
+        updates = {}
+
+        def flatten(o):
+            for k, v in o.items():
+                if k == "":
+                    inner_prefix = f"{ETCD_CONTAINER_REGISTRY_KEY}/{hostname}"
+                else:
+                    inner_prefix = f"{ETCD_CONTAINER_REGISTRY_KEY}/{hostname}/{k}"
+                if isinstance(v, Mapping):
+                    flatten(v)
+                else:
+                    updates[inner_prefix] = v
+
+        flatten(config)
```

Comment:
Even with inner functions, it's better to minimize the side-effects, particularly when the inner function mutates outer scope's values. Could we rewrite `flatten()` function to create and return `updates`? (The exception is when you write test codes to inspect inner activities.)